File: plugins/collector/event/apps/database/prometheus.go
package database

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

type Prometheus struct {
	version string
	reg     *regexp.Regexp
}

func (Prometheus) Name() string { return "prometheus" }

func (Prometheus) Type() string { return "database" }

func (p Prometheus) Version() string { return p.version }

func (Prometheus) Match(p *process.Process) bool { return p.Name == "prometheus" }

// Elkeid, failed to install in my ubuntu :-(
func (p *Prometheus) Run(proc *process.Process) (mapping map[string]string, err error) {
	if p.reg == nil {
		p.reg = regexp.MustCompile(`prometheus,\sversion\s(\d+\.)+\d+`)
	}
	result, err := apps.Execute(proc, "--version")
	if err != nil {
		return nil, err
	}
	str := p.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	p.version = strings.TrimPrefix(str, "prometheus, version ")
	return
}

func init() {
	apps.Regist(&Prometheus{})
}

File: plugins/collector/src/event/apps/database/prometheus.rs
use crate::event::apps::{execute, AppProc, IApp};
use std::collections::HashMap;

pub struct Prometheus {
    version: String,
}

impl Prometheus {
    pub fn new() -> Self { Prometheus { version: String::new() } }
}

impl IApp for Prometheus {
    fn name(&self)     -> &'static str { "prometheus" }
    fn app_type(&self) -> &'static str { "database" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "prometheus" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["--version"])?;
        // "prometheus, version 2.38.0 (branch: …)"
        for line in out.lines() {
            if let Some(rest) = line.strip_prefix("prometheus, version ") {
                self.version = rest.split_whitespace().next().unwrap_or("").to_owned();
                return Ok(HashMap::new());
            }
        }
        Err(anyhow::anyhow!("version not found"))
    }
}
